rust/operator-binary/src/reporting_task/mod.rs

```diff
@@ -0,0 +1,344 @@
+/// This will return
+/// * a Job that creates and runs the reporting task via the NiFi Rest API.
+/// * a Service that contains of one single NiFi node.
+///
+/// The Service is required in order to communicate only with one designated NiFi node.
+/// This is necessary as the generated JWT was changed in 1.25.0 and corrected the issuer
+/// from SingleUserLoginIdentityProvider to the FQDN of the pod.
+/// The NiFi role service will randomly delegate to different NiFi nodes which will
+/// then fail requests to other nodes.
+pub fn build_reporting_task(
+    nifi: &NifiCluster,
+    resolved_product_image: &ResolvedProductImage,
+    nifi_auth_config: &NifiAuthenticationConfig,
+    sa_name: &str,
+) -> Result<(Job, Service)> {
+    Ok((
+        build_reporting_task_job(nifi, resolved_product_image, nifi_auth_config, sa_name)?,
+        build_reporting_task_service(nifi, resolved_product_image)?,
+    ))
+}
+
+/// Return the name of the reporting task.
+pub fn build_reporting_task_service_name(nifi_cluster_name: &str) -> String {
+    format!("{nifi_cluster_name}-{REPORTING_TASK_CONTAINER_NAME}")
+}
+
+/// Return the FQDN (with namespace, domain) of the reporting task.
+pub fn build_reporting_task_fqdn_service_name(nifi: &NifiCluster) -> Result<String> {
+    let nifi_cluster_name = nifi.name_any();
+    let nifi_namespace: &str = &nifi.namespace().context(ObjectHasNoNamespaceSnafu)?;
+    let reporting_task_service_name = build_reporting_task_service_name(&nifi_cluster_name);
+
+    Ok(format!(
+        "{reporting_task_service_name}.{nifi_namespace}.svc.cluster.local"
+    ))
+}
+
+/// Return a `statefulset.kubernetes.io/pod-name` label for the first pod of the (sorted,BTreeMap)
+/// first rolegroup containing more than zero replicas. This is required to only select
+/// a single node in the Reporting Task Service.
+/// If no replicas are available or the cluster is stopped, no additional label will be returned.
+fn label_ordered_for_service_selector(nifi: &NifiCluster) -> BTreeMap<String, String> {
+    let cluster_name = nifi.name_any();
+    let node_name = NifiRole::Node.to_string();
+
+    // sort the rolegroups to avoid random sorting and therefore unnecessary reconciles
+    let sorted_role_groups = nifi
+        .spec
+        .nodes
+        .iter()
+        .flat_map(|role| &role.role_groups)
+        .collect::<BTreeMap<_, _>>();
+
+    let mut extra_service_label = BTreeMap::new();
+
+    // find the first rolegroup that has any replicas
+    for (role_group_name, role_group) in sorted_role_groups {
+        if let Some(replicas) = role_group.replicas {
+            if replicas > 0 {
+                extra_service_label.insert(
+                    "statefulset.kubernetes.io/pod-name".to_string(),
+                    format!("{cluster_name}-{node_name}-{role_group_name}-0"),
+                );
+                break;
+            }
+        }
+    }
+
+    extra_service_label
```

Comment:
In case extra_service_label is empty here (e.g. all replicas set to null to use HPAs), I think I would pick replica 0 of the first rolegroup as this is better than nothing

Reply:
> the service will be updated once the pods come back online.

I *think* here we are looking at the NifiCluster, not the actual state on the k8s cluster. IMHO we should support users not setting any replica number and instead deploying a HPA (as documented in https://docs.stackable.tech/home/stable/concepts/operations/#_performance, section "2. Performance")